src/annoylib.h

```diff
@@ -1172,7 +1206,87 @@ template<typename S, typename T, typename Distance, typename Random>
     return get_node_ptr<S, Node>(_nodes, _s, i);
   }
 
-  S _make_tree(const vector<S >& indices, bool is_root) {
+  void _thread_build(int q, int thread_idx, ThreadBarrier& barrier, std::mutex& _nodes_mutex) {
```

Comment:
i think it would be good to have the same code for threaded and non-threaded building

Reply:
I'm not exactly sure what you mean by this: should we still support building without threads (e.g. when n_threads=1)?

Reply:
I'd actually be ok only supporting threaded build so that we can remove the non-threaded build code

Reply:
I'm probably missing something, but `_thread_build` is just a private helper function that was extracted from the previous `build` function. It is there just so I don't have to stuck everything inside a lambda function when spawning a new thread.

Could you point out which part of the code could be potentially removed? :smile:

Reply:
I don't really understand why `_thread_build` is even needed? Couldn't you just call `_make_tree` and make sure each thread locks the part of the code where nodes are added? The onlyt two places are 
* https://github.com/spotify/annoy/blob/master/src/annoylib.h#L1185
* https://github.com/spotify/annoy/blob/master/src/annoylib.h#L1252

I don't think you need the magic where you go into nodes and add offsets etc? I could be wrong

Reply:
If it's not clear what I'm saying then give me a few days and I can try to get it working and then send a pull request to your pull request!

Reply:
That is how I did it in my initial implementation, but the problem lies in the `_allocate_size` function.

When multiple threads are building trees and one of them has to resize the memory (with `realloc`), it invalidates all pointers  held by other threads here: https://github.com/spotify/annoy/blob/master/src/annoylib.h#L1199
This leads to a segmentation fault. To guard against this, each thread would have to copy the children before using them. Additionally, all `_allocate_size` calls would have to be guarded by an exclusive mutex and all `_get` calls with a shared mutex (because we would have to ensure `_get` does not return an invalid pointer). This resulted in a significant slow-down due to mutex contention and increase in memory usage due to copying.

Reply:
Ok, that makes a lot of sense!

Instead of pointers to nodes, you could store their indices in the `children` array. That way the pointers won't get invalidated. However like you said, `_get` would still need to acquire a shared mutex. But I think instead of putting the mutex on `_get`, if you use your barrier approach, then aren't we guaranteed no thread will update the underlying node storage while it's being accessed? Not 100%

Reply:
Depends on where you would put the barrier. It sounds like it might work, but I'm worried this might significantly impact performance, because `_nodes` array is constantly accessed/written to. That is why I wanted to get rid of any critical sections in `_make_tree`, to make threads as independent as possible.

Reply:
When you said putting a mutex on `_get` resulted in a significant slow-down, was that during the tree building phase? Or during the search phase?

Reply:
Just the building phase. I put the lock around the parts you linked to (including the `_get` call).

Reply:
Another tricky thing might be support for `std::shared_mutex` (provides shared and exclusive locking capabilities), which is only supported from C++17 onward.

Other than that, I think your proposal could work. `create_split` would have to take the `_nodes` array as a parameter and a read lock would have to be placed around `create_split` and around the parts where `Node` pointers are active (returned from `_get`). And `_allocate_size` would have to protected with an exclusive lock.

Reply:
Yeah exactly, that's what i was thinking.

Your approach at https://github.com/spotify/annoy/pull/495/files#diff-91eaadc9d0aa248375a22a64f315e3d8R1250 is very clever but i'm a bit nervous it's too hacky and I'd love to have something slightly simpler :)

Doesn't your barrier thing roughly accomplish what at shared_mutex does? Or it should at least be possible to modify it into sort of a shared mutex? Haven't thought about it enough.

Reply:
I agree, I would definitely feel more comfortable if I could simplify it a bit :smile: 

Unfortunately a barrier does not accomplish what a shared_mutex is supposed to do. It can be only used as a global synchronization point, not for reader/writer style locking.

After reading a couple of SO threads, a custom shared mutex could prove really tricky to implement correctly. We could use the `shared_timed_mutex` which has a few additional capabilities and is available from C++14. Either way, I'll implement the less-magic version we discussed and then we can compare the performance and decide if it's worth it to bump the C++ version.

Reply:
Well, I managed to simplify my solution in the latest commit :smile: I removed all of the magic and introduced a couple of more `mutex` variables and a `shared_timed_mutex` variable.

I've re-ran my time measurements on glove-angular-25:
| Num. threads | n_trees=32 time (s) | n_trees=64 time (s) |
|--------------|---------------------|---------------------|
|            1 |  49.286592960357666 | 99.88854050636292   |
|            2 |   26.54487133026123 | 53.29321575164795   |
|            4 |  15.026970863342285 | 29.342100143432617  |
|            8 |  10.422403812408447 | 18.966835498809814  |
|           16 |   9.757703065872192 | 17.069363117218018  |

Reply:
Wow amazing! Any idea if there's any slowdown compared to the current master?

The new code looks much cleaner and less "clever". I don't know how well supported c++14 is but hopefully it's not a big problem.

Do you really need so many different mutexes? Not a concern, was just surprised to see 4 different ones

Reply:
I can prepare a more comprehensive timing benchmark and include a comparison with the current master. But as far as I can tell, there isn't a significant slowdown.

Regarding the number of mutexes: If I guard everything with the shared `nodes_mutex` just one thread has to acquire an exclusive lock and everything grinds to a halt. So I opted for a more granular approach, where I exclusively lock the `nodes_mutex` only when absolutely necessary and add a separate regular locks for all other shared variables (`nodes_size`, `_n_nodes`). I'll see if I can cut down on the number of mutexes, but I currently can't see a better way.